vault/core.go

```diff
@@ -830,6 +833,51 @@ func (c *Core) Unseal(key []byte) (bool, error) {
 		c.logger.Info("core: vault is unsealed")
 	}
 
+	// Fetch the metadata associated with the unseal keys
+	entry, err := c.barrier.Get(coreUnsealMetadataPath)
+	if err != nil {
+		return false, fmt.Errorf("failed to fetch unseal metadata entry: %v", err)
+	}
+
+	// For BC compatibility, log the metadata information only if it is
+	// available
+	var unsealMetadataEntry unsealMetadataStorageEntry
+	if entry != nil {
+		// Decode the unseal metadata information
+		if err = jsonutil.DecodeJSON(entry.Value, &unsealMetadataEntry); err != nil {
+			return false, fmt.Errorf("failed to decode unseal metadata entry: %v", err)
+		}
+
+		for _, unlockPart := range c.unlockParts {
+			// Fetch the metadata associated to the unseal key shard
+			keyMetadata, ok := unsealMetadataEntry.Data[base64.StdEncoding.EncodeToString(unlockPart)]
```

Comment:
This feels dirty, we should never store the unseal keys (in the barrier or not). Let's instead store a SHA256 of them or something.